### PR TITLE
Add suspicious_operation_groupings lint

### LINKED ISSUE #6039: Wrong similar code lint
This is a very common type of bugs that I'd like Clippy to catch (clippy::correctness), I think it's too much to ask rustc to do this:

```rust
struct Vec3 { x: f64, y: f64, z: f64 }

impl Eq for Vec3 {}

impl PartialEq for Vec3 {
    fn eq(&self, other: &Self) -> bool {
        self.x == other.y && self.y == other.y && self.z == other.z
    }
}

fn main() {}
```

For C++ similar to that Rust code PVS-Studio gives the error:

V1013 Suspicious subexpression a.x == b.y in a sequence of similar comparisons. vec3.h 182

I predict that if you implement a similar lint in a reasonable way then people that use Clippy regularly will suddenly find several new bugs in their code :-)

### PR DESCRIPTION
**Pull request by Ryan1729**
This is my (<del> currently WIP </del>) attempt to close #6039.

changelog: Added `suspicious_operation_groupings` lint.

### CODE REVIEW

**Comment by Ryan1729** [comment]
So at the moment several portions of the lint code are incomplete and only compile via the `todo` macro.

The main reason I'm submitting this as a draft PR intiialy is that as part of this lint  I would like to provide a suggestion that swaps a dynamically discovered identifier in an expression with another identifier. Creating a suggestion by changing an identifier in an expression seems like the kind of things that might have already been needed elsewhere. But after searching for a while in the clippy codebase I wasn't able to find a previous example of that.

Before attempting to create a way to do that, I'd like to know whether I've missed someplace where identifier swapping, or something similar has already been done. Alternatively, I'd also be interested in ideas anyone has on how to implement identifier swapping, given it has not been implemented already.

My best guess at how to implement it would be to recurse down the AST and find the `Span` of the identifier in question, then use that to create a `Span` with the previous part of the surrounding expression, and another `Span` with the part after that, then combine that with `Span` for the desired new identifier to make a suggestion string. For whatever reason I find myself unsure of that approach, but if I don't hear any objections I suppose I'll try it.

**Comment by Ryan1729** [comment]
After taking the time to do a more through search of the clippy code base, I have concluded that the most similar thing to what I need here that already exists is [`mirroed_exprs`](https://github.com/rust-lang/rust-clippy/blob/daad592fac4b9f582f0dfeb1c0ca5874b2a0b111/clippy_lints/src/unnecessary_sort_by.rs#L67-L170) which, while related, is solving a different problem that doesn't really fit into what I would want here, since it expects you to know the idents beforehand.

**Comment by ebroto** [comment]
Hey @Ryan1729, I've seen there is another WIP PR that this one depends on, and that you plan to rework that PR using a visitor. Feel free to ping me when you think this one is ready for review :)

File: tests/ui/suspicious_operation_groupings.stderr [hunk 1]
```diff
@@ -6,11 +6,23 @@ LL |         self.x == other.y && self.y == other.y && self.z == other.z
    |
    = note: `-D clippy::suspicious-operation-groupings` implied by `-D warnings`
 
+error: This sequence of operators looks suspiciously like a bug.
+  --> $DIR/suspicious_operation_groupings.rs:11:9
+   |
+LL |         self.x == other.y && self.y == other.y && self.z == other.z
```
Comment (Ryan1729) [inline tests/ui/suspicious_operation_groupings.stderr:12]: This lint message is emitted twice, and my best guess as to why, is that both `self.x == other.y && self.y == other.y && self.z == other.z` and `self.x == other.y && self.y == other.y` have detectable errors, and they both get visited separately by the lint pass. 

I'm unaware of a good way to prevent these double messages. Is this something that has been addressed for other lints before?

Comment (ebroto) [reply tests/ui/suspicious_operation_groupings.stderr:12]: I can't remember now a similar case. I think you can leave it like this for now and once we get to review we can try to find a solution.

**Comment by Ryan1729** [comment]
@ebroto As far as I understand the current CI failure is due to a change in `rustc` so I think this ready for review.

**Comment by ebroto** [comment]
r? @llogiq I'm a bit overwhelmed by PRs and quite short on time, would you mind taking over this one? Feel free to reassign it to me if you can't!

I think once this is ready we should remember to squash as there are many intermediate commits.

**Comment by llogiq** [comment]
Sure thing, @ebroto!

**Comment by llogiq** [comment]
Whoa! I'll need more than a few minutes to review this one...more like a long weekend...

**Comment by Ryan1729** [comment]
Yeah, I can tell that this PR has gotten large. That’s why I decided to postpone a feature and make issue #6275 about getting to that later.

Would it be useful to squash commits containing consecutive rewrites of the same parts of the code, now, before review?

Is there anything I can do to make reviewing this PR easier? Splitting the PR into smaller PRs maybe?

**Comment by llogiq** [comment]
No, it's all good, I'll just need more time. Give me about two weeks.

File: clippy_lints/src/utils/ast_utils.rs [hunk 2]
```diff
@@ -10,6 +10,28 @@ use rustc_ast::{self as ast, *};
 use rustc_span::symbol::Ident;
 use std::mem;
 
+pub mod ident_iter;
+pub use ident_iter::IdentIter;
+
+pub fn is_useless_with_eq_exprs(kind: BinOpKind) -> bool {
```
Comment (llogiq) [inline clippy_lints/src/utils/ast_utils.rs:16]: This could probably benefit from a local `use BinOpKind::*;`.

File: tests/ui/eq_op.rs [hunk 3]
```diff
@@ -86,3 +86,12 @@ fn check_ignore_macro() {
     // checks if the lint ignores macros with `!` operator
     !bool_macro!(1) && !bool_macro!("");
 }
+
+struct Nested {
+    inner: ((i32,), (i32,), (i32,)),
+}
+
+fn check_nested(n1: &Nested, n2: &Nested) -> bool {
+    // There's no `n2.inner.0.0`
```
Comment (llogiq) [inline tests/ui/eq_op.rs:95]: I was confused when first reading this comment. Maybe reword it as follows:

```suggestion
    // `n2.inner.0.0` mistyped as `n1.inner.0.0`
```

File: clippy_lints/src/utils/ast_utils.rs (hunk 2 above)
Comment (Ryan1729) [reply clippy_lints/src/utils/ast_utils.rs:16]: Agreed.

**Review by llogiq — COMMENTED** [review]
First round of review, no blockers so far but a few suggestions.

File: clippy_lints/src/suspicious_operation_groupings.rs [hunk 4]
```diff
@@ -0,0 +1,693 @@
...
+    /// // same as above except:
+    /// impl PartialEq for Vec3 {
+    ///     fn eq(&self, other: &Self) -> bool {
+    ///         // Note we now compare other.x to self.x
+    ///         self.x == other.x && self.y == other.y && self.z == other.z
+    ///     }
+    /// }
+    /// ```
+    pub SUSPICIOUS_OPERATION_GROUPINGS,
+    correctness,
```
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/suspicious_operation_groupings.rs:61]: `suspicious_arithmetic_impl` is a `complexity` lint, but I guess that triggers on a much smaller set of possible code. If it turned out that, in practice, the accuracy of this lint was sufficiently high, (that is, if no one ever actually needed to do the operations this lint flags,) then I would argue that `correctness` would still make sense. But admittedly, we don't know how accurate this lint it in practice. 

The unfortunate thing is that if the lint doesn't at least warn by default, then fewer people will end up using it, and we will have less data about how often people really want to do operations that this lint flags as suspicious. I think `style` is also a good choice since if the lint does turn out to have false positives, then (I assume) people would be more comfortable disabling it.

So given all that I'd rather change it to `style`, but if there is strong opposition to that, then changing it to `pedantic` makes sense tome too.

File: tests/ui/suspicious_operation_groupings.stderr (hunk 1 above)
Comment (Ryan1729) [reply tests/ui/suspicious_operation_groupings.stderr:12]: How do we prevent the sub-expressions from triggering the lint again after the parent expression collects all the suggestions after that? As far as I can tell `check_expr` gets called for each `expr` and each `expr` contained inside that `expr`.

File: clippy_lints/src/suspicious_operation_groupings.rs [hunk 5]
```diff
@@ -0,0 +1,693 @@
...
+            let mut op_types = Vec::with_capacity(binops.len());
+            // We could use a hashmap, etc. to avoid being O(n*m) here, but
+            // we want the lints to be emitted in a consistent order. Besides,
+            // m, (the number of distinct `BinOpKind`s in `binops`)
+            // will often be small, and does hav an upper limit.
+            for b in &binops {
+                if !op_types.contains(&b.op) {
+                    op_types.push(b.op);
+                }
+            }
```
[the remark this answers is not in the document]
Comment (Ryan1729) [reply clippy_lints/src/suspicious_operation_groupings.rs]: Turns out this doesn't borrow check, since `op_types` is in two closures. So I combined the `filter` and `for_each` into a single `for_each` instead.

File: clippy_lints/src/suspicious_operation_groupings.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (llogiq) [reply clippy_lints/src/suspicious_operation_groupings.rs]: Ok, then let's make it a `style` lint for now. We really need another category for lints like this.

**Comment by llogiq** [comment]
r=me with lint group changed to `style`.

**Comment by Ryan1729** [comment]
@llogiq I've squashed some of the commits as @ebroto [suggested above](https://github.com/rust-lang/rust-clippy/pull/6086#issuecomment-719904396), and the pushed commits now include [making the lint a style lint](https://github.com/rust-lang/rust-clippy/pull/6086/commits/216600683715857bdc017b4d5bb566f092536bfa).

**Comment by llogiq** [comment]
I think @ebroto meant to squash all the commits. Otherwise great job! :+1:

**Comment by Ryan1729** [comment]
Hmm. If I look at the recent history on this repo I see series of multiple commits that seem to come from the same PR. @ebroto , can you clarify what you wanted? I was under the impression that we only wanted to squash commits containing code that was replaced later by other code.

**Comment by Ryan1729** [comment]
@llogiq I decided to just go ahead and squash all the commits, since it made fixing an upstream merge conflict easier.

@rustbot modify labels: +S-waiting-on-review -S-waiting-on-author

**Comment by llogiq** [comment]
Sorry for the delays. Thank you for seeing this through with us. @bors r+
